Gun skill effects

comma opened the pull request:
Accuracy buffs, ignoring some of pain on higher effects, less info about gun shown to untrained, less reliable aimbot mode, more spread in bursts, sometimes randomly turning safety on (safety is now a thing)

Safety is now explicit, so help intent doesn't prevent discharge. It's is VERY important to check safety manually if you're unskilled, anything with basic skil or more do it automatically more or less, so no embarassing accidental discharges.

Professional skill gets bonuses to handling sniper rifles more accuracy and automatic ejection of shell in bolt action one.

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

babydoll commented:
Alt click will conflict with several alt-clicks on guns, most notably revolvers.

Also somewhat adverse to giving bonuses here, but will need to read the code

ghost commented:
You gotta set it to professional by default for antags for that sexy autoejection.

comma commented:
Made it ctrl-click (when carried somewhere on mob) and rightclick verb just in case

ghost commented:
Can you also reduce the awful recoil for those who have higher weapons training?

comma commented:
Eh, its just cosmetics, and barely a thing anyway. Miggt look at ut but  unlikely

EcklesFire commented:
It feels a bit weird that the safety randomly turns itself off.

Ehhhh.

HeyBanditoz commented:
New PR: removes intents.

On June 10, 2018 5:51:53 AM MDT, Chinsky <notifications@github.com> wrote:
>Accuracy buffs, ignoring some of pain on higher effects, less info
>about gun shown to untrained, less reliable aimbot mode, more spread in
>bursts, sometimes randomly turning safety on (safety is now a thing)
>
>Safety is now explicit, so help intent doesn't prevent discharge. It's
>is VERY important to check safety manually if you're unskilled,
>anything with basic skil or more do it automatically more or less, so
>no embarassing accidental discharges.
>
>Professional skill gets bonuses to handling sniper rifles more accuracy
>and automatic ejection of shell in bolt action one.
>
><!-- 
>Do not forget to add a changelog when you have made admin/player facing
>changes that can alter gameplay.
>Examples which require a changelog entry include:
>* Adding/removing objects that players may interact with, or the way
>they function.
>* Adding/removing/altering admin tools.
>* Changing the map.
>
>Examples were changelog entries are optional/not typically required:
>* Cosmetic changes such as descriptions, sound effects, etc.
>* Optimizations and other changes to underlying systems which do not
>affect gameplay.
>* Minor bug fixes.
>
>You find a README and example file in .\html\changelogs\ for further
>instructions.
>-->
>
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/Baystation12/Baystation12/pull/21749
>
>-- Commit Summary --
>
>  * Gun skill effects
>
>-- File Changes --
>
>    M code/modules/mob/living/carbon/human/human.dm (10)
>    M code/modules/mob/skills/skill.dm (19)
>    M code/modules/projectiles/gun.dm (51)
>    M code/modules/projectiles/guns/projectile.dm (13)
>    M code/modules/projectiles/guns/projectile/sniper.dm (18)
>    M code/modules/projectiles/targeting/targeting_triggers.dm (7)
>    A html/changelogs/chisnky - gungungun.yml (10)
>    M maps/torch/job/jobs.dm (2)
>
>-- Patch Links --
>
>https://github.com/Baystation12/Baystation12/pull/21749.patch
>https://github.com/Baystation12/Baystation12/pull/21749.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/Baystation12/Baystation12/pull/21749

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

HeyBanditoz commented on code/modules/projectiles/gun.dm:
```diff
@@ -426,9 +450,12 @@
+		if(firemodes.len > 1)
+			var/datum/firemode/current_mode = firemodes[sel_mode]
+			to_chat(user, "The fire selector is set to [current_mode.name].")
+	to_chat(user, "Safety is [safety ? "on" : "off"].")
```
The safety...

HeyBanditoz commented on code/modules/projectiles/gun.dm:
```diff
@@ -451,6 +478,26 @@
+/obj/item/weapon/gun/proc/toggle_safety(var/mob/user)
+	if(loc == user)
+		safety = !safety
+		to_chat(user, "<span class='notice'>You switch safety [safety ? "on" : "off"] on [src].</span>")
```
switch the safety...

HeyBanditoz commented on code/modules/projectiles/guns/projectile/sniper.dm:
```diff
@@ -26,15 +26,27 @@
 
+/obj/item/weapon/gun/projectile/heavysniper/handle_post_fire(mob/user, atom/target, var/pointblank=0, var/reflex=0)
+	if(user && user.skill_check(SKILL_WEAPONS, SKILL_PROF))
+		to_chat(user, "<span class='notice'>You work the bolt open with a reflexive motion, ejecting [chambered]!</span>")
```
Alternatively: "You reflexively open the bolt, ejecting..."

HeyBanditoz commented on code/modules/projectiles/targeting/targeting_triggers.dm:
```diff
@@ -19,10 +19,11 @@
-	if(owner.a_intent == I_HELP)
-		to_chat(owner, "<span class='warning'>You refrain from firing \the [aiming_with] as your intent is set to help.</span>")
+	if(prob(owner.skill_fail_chance(SKILL_WEAPONS, 30, SKILL_ADEPT, 3)))
+		to_chat(owner, "<span class='warning'>You fumble with your gun, throwing aim off!</span>")
```
Probably better to say the gun instead of your gun. Also, "...throwing your aim off."

HeyBanditoz reviewed (changes requested):
Grammar reeee

AtomicWorm commented:
Cue everybody on the Torch having professional-level firearms skill.

comma commented:
>It feels a bit weird that the safety randomly turns itself off.

It doesnt, it can randomly tyrn ON

>Cue everybody on the Torch having professional-level firearms skill.

Pretty expensive now, cists were bumped, higher tiers are as expensive as med skills

AtomicWorm commented:
>Pretty expensive now, cists were bumped, higher tiers are as expensive as med skills

Which adds another problem in that if anybody in a department that uses guns, such as security, wants to be good at it, they're gonna be no good for anything that _isn't_ using guns.

Hubblenaut commented:
I think the change of intent was a better control for gun safety. Could make people with a low skill level in ranged weaponry be able to fire the gun even on help intent, so they aren't safe from accidentally firing the gun in any intent.

Otherwise, everyone is likely to keep gun safety off at all times anyway. Also, I don't think CTRL + Click is an intuitive shortcut as it's not used for anything safe inflatables - and it only exists for inflatables because people were vehemently against ALT + Click in the PR for that.

comma commented:
>Which adds another problem in that if anybody in a department that uses guns, such as security, wants to be good at it, they're gonna be no good for anything that isn't using guns.

Which adds a problem of people using hyperbole when talking about things they didn't even check.
Run of mill MAA can easily get to Experienced and have skills to spare over other skills, like for dat Experienced close combat and althetics AND eva.
Now to get Professional you indeed would need to concentrate fairly heavily on gun skill, but that's the intent.

>be able to fire the gun even on help intent,

Everyone can fire on help intent now, since safety is an actual existing thing now.

> Also, I don't think CTRL + Click is an intuitive shortcut as it's not used for anything safe inflatables

Click is taken, Alt click is taken, worst case there's rightclick verb.

Hubblenaut commented:
I should have been clearer, but how about this:

Make the changing of intents trigger the safety, as whether or not you want to shoot the gun is directly correlative to the intent. A weapon that is drawn or picked up while not on help intent would also immediately have its safety flicked off, ready to shoot.

comma commented:
Eh, in my experience help intent thing was more of an annoyance and gotcha than actually preventing anything. You have to pick either not shoving people around, or actually being able to use gun.

chaoko99 commented:
>It doesnt, it can randomly tyrn ON

I'm sorry but that's legitimately worse.

comma commented:
>I'm sorry but that's legitimately worse.

That's the intent. Either get training or check your safety manually, after each examine/safety toggle you have some grace period where it won't happen.

babydoll commented:
@comma How hard would it be to draw a 'S' icon over the gun in your inventory if the safety is enabled? Then at least you can actually 'check' relatively quickly at a glance without having to examine.

comma commented:
could be kinda tricky to keep it to 'in inventory', and being able to tell state of safety of a gun lying on other end of the room sounds eh.

Hubblenaut commented:
Syringes do the same thing, they only make use of `pickup()` and `dropped()`.

comma commented:
Thanks, stole code from syringes.
Also makeshift weapons (zipguns, crossbows) now have no safety inbuilt.

MistakeNot4892 commented on code/modules/mob/skills/skill.dm:
```diff
@@ -192,11 +192,22 @@ GLOBAL_LIST_EMPTY(skills)
 	name = "Weapons Expertise"
 	desc = "This skill describes your expertise with and knowledge of weapons. A low level in this skill implies knowledge of simple weapons, for example flashes. A high level in this skill implies knowledge of complex weapons, such as unconfigured grenades, riot shields, pulse rifles or bombs. A low-medium level in this skill is typical for security officers, a high level of this skill is typical f ...
-	levels = list( "Unskilled"			= "You know how to recognize a weapon when you see one. You can probably use pepper spray or a flash, though you might fumble and turn them on yourself by mistake. You're likely to shoot yourself in the foot or forget to take the safety off. Your lack of training may make you more dangerous to your allies than your enemies.",
+	levels = list( "Unskilled"			= "You know how to recognize a weapon when you see one. You can point gun and shoot it, though results vary wildly. You might forget the safety, you can't control burst recoil well, and you don't have trained reflexes for gun fighting.",
```
`point a gun`

MistakeNot4892 commented on code/modules/projectiles/gun.dm:
```diff
@@ -161,6 +167,17 @@
+			to_chat(user, "<span class='warning'>[src] fires on its own!</span>")
+			var/list/targets = trange(2, src)
+			targets += user
+			if(LAZYLEN(targets))
```
Why are you checking len on targets immediately after adding the user?

comma replied:
don't trust anyone even yourself

Techhead0 commented on code/modules/projectiles/guns/projectile/sniper.dm:
```diff
@@ -26,15 +26,27 @@
+		if(!bolt_open)
+			playsound(src.loc, 'sound/weapons/flipblade.ogg', 50, 1)
+			bolt_open = 1
+		chambered.loc = get_turf(src)
```
ForceMove?
(Or `DropInto(user.loc)`)

PurpleMartinJCK commented on code/modules/projectiles/guns/launcher/crossbow.dm:
```diff
@@ -64,6 +64,12 @@
 	var/obj/item/weapon/cell/cell = null    // Used for firing superheated rods.
 	var/current_user                        // Used to check if the crossbow has changed hands since being drawn.
 
+/obj/item/weapon/gun/launcher/crossbow/safety()
```
This should be moved to a more general "has safety" var for guns that's checked in gun/safety

Miraviel commented:
> @comma How hard would it be to draw a 'S' icon over the gun in your inventory if the safety is enabled? Then at least you can actually 'check' relatively quickly at a glance without having to examine.

How about a quiet clicking sound whenever you turn the safety on/off? Something similar to the no-ammo clicking noise but deeper. It'd be an easier way to identify it, I always found paying attention to sounds easier than to sprites in ss13.

comma commented:
Already does yeah

comma committed:
Gun skill effects

Accuracy buffs, ignoring some of pain on higher effects, less info about gun shown to untrained, less reliable aimbot mode, more spread in bursts, sometimes randomly turning safety on (safety is now a thing)

Safety is now explicit, so help intent doesn't prevent discharge. It's is VERY important to check safety manually if you're unskilled, anything with basic skil or more do it automatically more or less, so no embarassing accidental discharges.

Professional skill gets bonuses to handling sniper rifles more accuracy and automatic ejection of shell in bolt action one.
